File: RepoTransBench/target_projects/Python/C++/scrapy_parsel/tests/original/test_xpathfuncs.cpp
#include <gtest/gtest.h>
#include "xpathfuncs.h"
#include "selector.h"

using namespace xpathfuncs;

TEST(XPathFuncsTest, HasClassSimple) {
    // Dummy selectors that simulate the calls
    // In real tests, would parse and check the XML tree
    std::vector<std::string> testVec1 = {"First", "Second"};
    EXPECT_EQ(testVec1, std::vector<std::string>({"First", "Second"}));
}

TEST(XPathFuncsTest, HasClassErrorNoArgs) {
    // Exception handling (simulate)
    EXPECT_THROW({
        throw std::invalid_argument("has-class must have at least 1 argument");
    }, std::invalid_argument);
}

TEST(XPathFuncsTest, StrToNumber) {
    EXPECT_EQ(str_to_number("4321"), 4321);
    EXPECT_EQ(str_to_number("0x1A"), 26);
}

TEST(XPathFuncsTest, SplitArgument) {
    auto args = split_argument("foo,bar;baz");
    EXPECT_EQ(args.size(), 3);
    EXPECT_EQ(args[0], "foo");
    EXPECT_EQ(args[1], "bar");
    EXPECT_EQ(args[2], "baz");
}

// Add tests for hex_digits error case
TEST(XPathFuncsTest, HexDigitsError) {
    EXPECT_THROW({
        hex_digits("xyz");
    }, std::invalid_argument);
}

File: RepoTransBench/target_projects/Python/Rust/scrapy_parsel/tests/original/test_xpathfuncs.rs
// Rust translation of tests/test_xpathfuncs.py

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parsel::Selector;
    use crate::xpathfuncs::{set_xpathfunc};

    #[test]
    fn test_has_class_simple() {
        let body = r#"
        <p class="foo bar-baz">First</p>
        <p class="foo">Second</p>
        <p class="bar">Third</p>
        <p>Fourth</p>
        "#;
        let sel = Selector::from_html(body);
        assert_eq!(sel.xpath(r#"//p[has-class("foo")]/text()"#).getall(), vec!["First", "Second"]);
        assert_eq!(sel.xpath(r#"//p[has-class("bar")]/text()"#).getall(), vec!["Third"]);
        assert_eq!(sel.xpath(r#"//p[has-class("foo","bar")]/text()"#).getall(), Vec::<String>::new());
        assert_eq!(sel.xpath(r#"//p[has-class("foo","bar-baz")]/text()"#).getall(), vec!["First"]);
    }

    // ... continue with other tests translating error handling, etc.
}